### ISSUE TITLE
Create DB read/write connection pools for storagenodes

### ISSUE THREAD

**Issue by dlamarmorgan** [storj/storj#7010]
To help alleviate DB contention in the storagenodes, create separate connection pools for reads and writes to the SQLite DB. This has the additional benefit of giving more control over how read and write operations function to the application.

Additional context can be found at https://archive.is/Xfjh6

**Comment by storjrobot** [comment]
This issue has been mentioned on **Storj Community Forum (official)**. There might be relevant details there:

https://forum.storj.io/t/suggestion-to-have-sqlite-optimized-or-tailored/26730/1

**Comment by MaxMaximus** [comment]
I double down on this enhancement suggestion!

Separate database connections for writing (modification) and for reading requests are a "must-have" feature for proper work of almost any DB under a significant load.
Also, such separate write connections should preferably be closed periodically (on regular basis - actual interval is based on actual amounts of write/modify request to opened db), whereas read-only connections usually can remain open for a long time. Which allow SQLite to regularly process all accumulated transaction in write-ahead journal (`*.-WAL`) and persistent all pending tx to main database (`*.db`). 
Working with a long unprocessed journal (large `*.db-wal` file size) slows down the work of VERY much. Dozens or even hundreds of times. In a worst case scenarios it can be up to few thousand times (db files stored on HDD and -WAL size too big to fit in caches).

As far as I understand the architecture of SQLite working in a -WAL mode, this is happening due to the fact that with any request to the database, it not only reads/searches for the requested rows in the main database (which, with the correct assignment of indexes, is processed **very** quickly), but also scans the **whole** -WAL journal from beginning to the end in search of pending modifying(update/delete/etc) transactions that may affect the requested table rows.
If hundreds of thousands or even millions of pending transactions accumulate there in -WAL, this can slow down the work not just by several times, but by several orders of magnitude.

And this may already be happening when the TTL collector is running with a lot of already expired pieces in db, in which there is an extreme slowdown of its operation on large nodes, as described in the other issues there: https://github.com/storj/storj/issues/7042#issuecomment-2236329037 and there: https://github.com/storj/storj/issues/7053
Not sure about @littleskunk (please check it too) , but in my case on my large node with extreme slow work of TTL collector i already saw `piece_expiration.db-wal`  grow up to 3000-3500 MB it size. Although by default it should usually not exceed 4 MB (1000 standard db pages) according to SQLite docs. But only if DB works properly.

It is also useful to enable own SQLite RAM cache for at least a few dozen of MB if this has not already been done. And it's best to make it a configurable option, readable from the config at startup as SQLite allow to modify it at run time.

It's quite surprising to me to watch you create more and more different custom caches to get around database problems instead of solving them. As programmers often say in my country, "put more and more crutches". Instead of fixing a broken leg.
It comes to the ridiculous sometimes, recently, for example, you added a separate cache for writing updates to the database "`garbage_collection_filewalker_progress.db`" 
In which, when the GC is running, only exactly one line is updated (more precisely, even one value, eg one column of one row) and 1024 times for the entire GC pass, which takes from minutes at best to tens of hours.

For any normally functioning database, 1024 updates of a single row is simply **nothing**. This should not cause any problems and does not require "reinventing the wheel" with additional custom caches on top of it. 
Separate caches can only make sense for very frequently changed values. Such as traffic and space usage counters, which are updated with each processed piece. But they can be written to the database quite often too, not with each update but often (and not just once an hour as now). For everything else, it is better to keep it as simple as possible and generally get rid of additional levels, which periodically create additional problems in themselves (as has already happened in several cases of non-preservation of progress). And just update directly in DB. They are generally designed just for this exact purpose - reliability and speed of storage of frequently used data. But it need to be fixed first.

**Comment by MaxMaximus** [comment]
My knowledge and skills regarding databases are already very outdated (about 10-15 years) and my be no longer relevant here.
But almost 15 years ago, in a relatively small online project led by only two amateur programmers (jоunior in modern terminology, although we were not called that yet back ago), we had no any problems/errors getting stable database operation with a stream of transactions (a mix of reading and writing in relatively close proportions) of the order of several dozen tx per second. Only after a little tuning of the settings and a few performance tests.
Using a slow interpreted language (PHP), a simple DBMS (very old versions of MySQL) and very weak and crappy hardware by modern standards. It wasn't even hardware, but cheap virtual servers: about 1-2 virtual cores running at 0.5-1 GHz and 256-512 MB of RAM, and a regular HDD as a only disk.
In modern conditions, loads of at least few hundreds of transactions per second should not create problems even on a weak HW setups. If they create it, it means that something is wrong with the settings or the scheme of working with the database.

**Comment by storjrobot** [comment]
This issue has been mentioned on **Storj Community Forum (official)**. There might be relevant details there:

https://forum.storj.io/t/huge-large-wal-file/27194/2
